### DISCUSSION TITLE
Openzeppelin Clones `clone` and `cloneDeterministic` fails

### DISCUSSION THREAD

**Discussion by waynehoover** [zkSync-Community-Hub/zksync-developers#91] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

When using the Open Zepplin clones library to deploy an [EIP 1167](https://eips.ethereum.org/EIPS/eip-1167) minimal proxy I get [an error](https://goerli.explorer.zksync.io/tx/0x712e3fed661824936fbdeac64097042f41b8bcbce6b6bda98f4b1d8043f8c756). I believe this is because `create` and `create2` must be [aware of the bytecode](https://era.zksync.io/docs/reference/architecture/differences-with-ethereum.html#create-create2) before deployment, but the compiler should be aware of the bytecode because it is hardcoded into the clones contract, so I'm not sure why it's failing.

### What is the specific issue or error you're encountering?

`execution reverted: ERC1167: create failed`

### Can you share the error messages or logs you're receiving, if any?

Here is the failed TX: https://goerli.explorer.zksync.io/tx/0x712e3fed661824936fbdeac64097042f41b8bcbce6b6bda98f4b1d8043f8c756

```
Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={"reason":"execution reverted: ERC1167: create failed","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0x92e38e60c9e4e8aC6AcB24dE02d4a48856D775d4","maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"maxFeePerGas":{"type":"BigNumber","hex":"0x77359400"},"to":"0x49f05f2646772b6D0E695550CA49D1e3fF8A17E7","value":{"type":"BigNumber","hex":"0x00"},"data":"0x70dfd40a000000000000000000000000651c7fbbac1fa3a8937377c4879ca99c11f1261c0000000000000000000000000000000000000000000000000000000064d48af00000000000000000000000000000000000000000000000000000000064d3d4200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000002439613562323837322d613264362d343866392d386364662d633866653730626163663333000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","type":2,"accessList":null},"error":{"code":3,"data":"0x08c379a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000016455243313136373a20637265617465206661696c656400000000000000000000"}}, tx={"data":"0x70dfd40a000000000000000000000000651c7fbbac1fa3a8937377c4879ca99c11f1261c0000000000000000000000000000000000000000000000000000000064d48af00000000000000000000000000000000000000000000000000000000064d3d4200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000002439613562323837322d613264362d343866392d386364662d633866653730626163663333000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","to":{},"from":"0x92e38e60c9e4e8aC6AcB24dE02d4a48856D775d4","value":{"type":"BigNumber","hex":"0x00"},"type":2,"maxFeePerGas":{"type":"BigNumber","hex":"0x77359400"},"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.7.0)
```

### Have you made any recent changes to the contract before encountering this issue?

No.

### Are there any external libraries or contracts that your contract interacts with?

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v4.9/contracts/proxy/Clones.sol

### Can you provide the relevant portions of your contract code where the issue is occurring?

https://github.com/rabbitholegg/quest-protocol/blob/642a83d0c2bd00aca6cb9875adecaef19ce23283/contracts/QuestFactory.sol#L326

### Have you tried to isolate the problem, and if so, what were the results?

Yes, it's in the OZ clones library. 

### What steps have you already taken to try to resolve the issue?

I have tried both `clone` and `cloneDeterministic`.

People have mentioned this issue a number of times in discord, but the only response that was given I can see is here: https://discord.com/channels/722409280497516566/1037142085368807444/1039697756371042434, which doesn't help.

**Comment by idea404** [comment, marked as answer]
Hi @waynehoover, the current version of the plugin does not support this. There are plans to implement this in future however but we can not say exactly when, sometime in the next half year looks likely.

**Reply by idea404** [reply]
To add to the above: the reason why this is not currently supported is because EIP 1167 is written directly in EVM bytecode, which is quite different from the bytecode that zkEVM operates on. As a result, it's currently not feasible to use EIP 1167 on zkSync's Era. To support this pattern, an entirely new "minimal" bytecode sequence would need to be created. 

Something you could do instead is deploy duplicated contracts (might even be more cost-effective than Ethereum but not entirely sure about this).

**Reply by sebastiantf** [reply]
Prolly worth adding into official docs if the fix / upgrade is not coming any soon

**Reply by sebastiantf** [reply]
https://docs.zksync.io/build/developer-reference/differences-with-ethereum.html#create-create2

**Comment by sebastiantf** [comment]
Linking related discussion: #97

**Comment by johnwesener** [comment]
The error "execution reverted: ERC1167: create failed" typically occurs when there's an issue with deploying the minimal proxy contract using EIP-1167. This could be due to various reasons, such as incorrect bytecode, insufficient gas, or a problem with the deployment process.

Here are a few steps you can take to troubleshoot and resolve this issue:

Verify the bytecode: Ensure that the bytecode of the minimal proxy contract is correct and matches the expected bytecode. Any discrepancy here could lead to deployment failures.

Gas estimation: Check if the gas limit provided for the deployment is sufficient. Sometimes, complex deployments require more gas than anticipated. You may need to manually specify a higher gas limit to ensure successful deployment.

Deployment process: Double-check the deployment process to ensure that all necessary parameters are correctly passed, such as constructor arguments and initialization steps. Any missing or incorrect parameter could cause deployment failures.

Test different environments: Try deploying the contract on different test networks or locally to see if the issue persists across multiple environments. This can help identify whether the problem is specific to a certain environment or configuration.

Debugging tools: Use debugging tools provided by your development environment or blockchain platform to get more insights into why the deployment is failing. This could include tools for analyzing transaction traces, gas consumption, and contract interactions.

By systematically investigating these areas, you should be able to identify the root cause of the deployment failure and take appropriate actions to resolve it.

**Reply by sebastiantf** [reply]
https://docs.zksync.io/build/developer-reference/differences-with-ethereum.html#create-create2

**Comment by natyfikapsha** [comment]
It seems like the issue might be related to the bytecode used for deployment. The error message "ERC1167: create failed" suggests that there could be a problem with the bytecode generation or deployment process. It's worth double-checking the bytecode and ensuring that it matches the expected format for EIP-1167 minimal proxies. Additionally, verifying that the necessary constructor parameters are being passed correctly during deployment could help resolve the issue.

**Reply by sebastiantf** [reply]
https://docs.zksync.io/build/developer-reference/differences-with-ethereum.html#create-create2

**Comment by sfatitetyccindy** [comment]
Issue: Unable to deploy EIP 1167 minimal proxy using OpenZeppelin Clones library.
Error: "execution reverted: ERC1167: create failed"
Possible Causes: Issue may be related to bytecode mismatch or gas estimation.
Steps Taken: Tried both clone and cloneDeterministic methods, but error persists.
Additional Info: Similar issues discussed in Discord, but no conclusive solution provided.

**Comment by latoyatidwell** [comment]
The issue seems to be related to gas estimation and the deployment of an EIP-1167 minimal proxy using OpenZeppelin's Clones.sol library. Further investigation into gas estimation and contract deployment settings may be needed to resolve this issue.
